infrastructure/fw-trace/fw-trace.py

```diff
@@ -0,0 +1,407 @@
+					continue
+				else:
+					print ('--->Ingress firewall rule "%s" matched.  Traffic is allowed to ingress to %s' % (item['name'],dst))
+					implicit_deny = False
+					break
+		
+			#check to see if protocol/port match denied list
+			if 'denied' in item:
+				denied_matched = False
+
+				for prot in item['denied']:
+					if prot['IPProtocol'] == "all":
+						if str(verbose) <> "None" : print ('  Ingress firewall rule "%s" allows all protocols and ports' % (item['name']))
+						denied_matched  = True
+						break
+				
+					matched_port = True
+					if ((prot['IPProtocol'] == "tcp") or (prot['IPProtocol'] == "udp")) and (prot['IPProtocol']  == protocol):
+						#checking to see if port matches any item is port list
+						matched_port = False
+						for p in prot['ports']:	
+							if p == port: 
+								matched_port = True
+								break
+							if '-' in p:				
+								if (int(port) >= int(p.partition('-')[0])) and (int(port) <= int(p.partition('-')[2])):
+									matched_port = True
+									break							
+														
+					if (protocol == prot['IPProtocol']) and matched_port :
+						if str(verbose) <> "None" : print ('  Ingress firewall rule "%s" with allowed protocols %s matches the protocol %s:%s  ' % (item['name'], item['allowed'],protocol,port))
+						denied_matched  = True
+						break   
+
+			
+				if not denied_matched :
+					if str(verbose) <> "None" : print ('  Ingress firewall rule "%s" does not allow or deny traffic on protocol %s:%s. Skipping to next firewall rule' % (item['name'], protocol, port))
+					continue
+				else:
+					print ('--->Ingress firewall rule "%s" matched.  Traffic ingress to %s is blocked' % (item['name'],dst))	
+					implicit_deny = False	
+					break
+					
+        if implicit_deny: print ('--->Reached end of ingress firewall rules. Traffic ingress to %s is implicitly blocked' % dst)	
+        print (' ')	       
+	
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+
+    parser.add_argument('src', help='Source IP')
+    parser.add_argument('dst', help='Destination IP')
+    parser.add_argument('protocol', help='Protocol')
+    parser.add_argument('port', help='Port')
+    
+    parser.add_argument(
+        '--project',
+        default='',
+        help='Your Google Cloud project name')
+        
+    parser.add_argument(
+        '--network',
+        default='default',
+        help='Your Google Cloud network name')
+
+    parser.add_argument(
+        '--src_tag',
+        help='Your Google Cloud firewall tag for the source of the traffic within GCP')
+
+    parser.add_argument(
+        '--dst_tag',
+        help='Your Google Cloud firewall tag for the destination of the traffic within GCP')
+        
+    parser.add_argument(
+        '--verbose',
```

Comment:
use  action='store_true'  as we don't want to supply an argument to the --verbose flag.